libvast/vast/system/terminator.hpp

```diff
@@ -0,0 +1,47 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include "vast/detail/type_traits.hpp"
+
+#include <caf/actor_addr.hpp>
+#include <caf/behavior.hpp>
+#include <caf/response_promise.hpp>
+#include <caf/stateful_actor.hpp>
+
+#include <vector>
+
+namespace vast::policy {
+
+struct sequential {};
+struct parallel {};
+
+} // namespace vast::policy
+
+namespace vast::system {
+
+struct terminator_state {
+  std::vector<caf::actor_addr> remaining_actors;
+  caf::response_promise promise;
+  static inline const char* name = "terminator";
+};
+
+template <class Policy>
+caf::behavior terminator(caf::stateful_actor<terminator_state>* self);
+
+/// @relates terminator
```

Comment:
From an API perspective, it feels a bit weird to me that the caller is supposed to create the terminator and then do nothing with it but pass it into this function.

Maybe we should just have functions `shutdown_sequential()` and `shutdown_generic()` (or `shutdown<Policy>()`) that create the terminator internally?

Also, I'm wondering if it would make sense to return the promise here, so the caller can wait for successful termination directly, without having to use `attach_functor()`.

Reply:
I've modularized the terminator actor a bit. There are now different helper utilities to interact with it:

- `shutdown`: non-blocking termination that also terminates the caller.
- `wait`: blocking termination modeled after `scoped_actor::wait_for`.

All the helper functions include spawning the `terminator`, so it's no longer necessary at the call site. Looks way more reasonable that way.